src/foundry/client/canvas/placeables/tokens/ruler.d.mts

```diff
@@ -1,13 +1,211 @@
+  ): TokenRuler.GridHighlightStyle;
 }
 
 declare namespace TokenRuler {
   interface Any extends AnyTokenRuler {}
   interface AnyConstructor extends Identity<typeof AnyTokenRuler> {}
+
+  interface Outline {
+    thickness: number;
+    color: PIXI.ColorSource;
+  }
+
+  interface DashLine {
+    dash: number;
+    gap: number;
+    speed: number;
+  }
+
+  interface Data {
+    /**
+     * The waypoints that were already passed by the Token
+     */
+    passedWaypoints: unknown[]; // TODO: Token.MeasuredMovementWaypoint
+
+    /**
+     * The waypoints that the Token will try move to next
+     */
+    pendingWaypoints: unknown[]; // TODO: Token.MeasuredMovementWaypoint
+
+    /**
+     * Movement planned by Users
+     */
+    plannedMovement: Record<string, unknown>; // TODO: Token.PlannedMovement
+  }
+
+  interface WaypointData {
+    /** The config of the movement action */
+    actionConfig: CONFIG.Token.MovementActionConfig;
+
+    /** The ID of movement, or null if planned movement. */
+    movementId: string | null;
+
+    /** The index of the waypoint, which is equal to the number of explicit waypoints from the first to this waypoint. */
+    index: number;
+
+    /** The stage this waypoint belongs to */
+    stage: "passed" | "pending" | "planned";
+
+    /** Is this waypoint hidden? */
+    hidden: boolean;
+
+    /** Is this waypoint unreachable? */
+    unreachable: boolean;
+
+    /** The center point of the Token at this waypoint. */
+    center: PIXI.Point;
+
+    /** The size of the Token in pixels at this waypoint. */
+    size: {
+      width: number;
+      height: number;
+    };
+
+    /** The ray from the center point of previous to the center point of this waypoint, or null if there is no previous waypoint. */
+    ray: foundry.canvas.geometry.Ray | null;
+
+    /** The measurements at this waypoint. */
+    measurement: foundry.grid.BaseGrid.MeasurePathResultWaypoint; // TODO: GridMeasurePathResultWaypoint
+
+    /** The previous waypoint, if any. */
+    previous: Waypoint | null;
+
+    /** The next waypoint, if any. */
+    next: Waypoint | null;
+  }
+
+  type Waypoint = DeepReadonly<Omit<TokenDocument.MeasuredMovementWaypoint, "movementId"> & WaypointData>;
+
+  /**
+   * @remarks Intended to be extended by subclasses that need to track additional info between waypoints
+   * Importantly, this is *not* saved to the token's movement data, instead it is merely mutated locally.
+   */
+  interface State {
+    hasElevation: boolean;
+    previousElevation: number;
+  }
+
+  /**
+   * @remarks Fed into the template
+   */
+  interface WaypointContext {
+    action: string;
+    cssClass: string;
+    secret: boolean;
+    units: string;
+    uiScale: number;
+    position: {
+      x: number;
+      y: number;
+    };
+    distance: SegmentDistance;
+    cost: SegmentCost;
+    elevation: ElevationContext;
+  }
+
+  interface SegmentDistance {
+    total: string;
+    delta?: string;
```

Comment:
Could you check for `| undefined`?

Reply:
So I want `| undefined` to be considered on output interfaces for 3 reasons:
1. I like rules like this to be applied as consistently as possible.
2. Users may override the relevant methods and find it limiting to be unable to write `delta: blah()`.
3. From my understanding Foundry puts little to no consideration into keeping this consistent across versions. They won't make a documentation change because they believe them to be interchangeable.

I really only have _any_ audit step for `undefined` because stuff like `mergeObject` and friends throw a wrench into just naively adding `| undefined`.